### PR TITLE
Reader support for encrypted DWRF files

### PR DESCRIPTION
**Pull request by rschlussel**
Support for reading encrypted columns in the DWRF file format. I still need to add tests. Also, this PR just contains the reader logic.

Here are the items left for full reader support. I may do them in follow up PRs 
* support configuring the decryption library
* get keys from key service in coordinator and pass through to reader.
* support configuring key service.

```
== RELEASE NOTES ==

Hive Changes
* TODO
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 1]
```diff
@@ -105,7 +109,41 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+    }
+
+    private KeyProvider toKeyProvider(DwrfProto.Encryption.KeyProvider keyProvider)
+    {
+        switch (keyProvider) {
+            case CRYPTO_SERVICE:
+                return KeyProvider.CRYPTO_SERVICE;
+            default:
+                // TODO: should we throw an exception instead?
+                return KeyProvider.UNKNOWN;
```
Comment (rschlussel) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:145]: @zzhao0  what's the correct behavior for UKNOWN key provider type?

Comment (zzhao0) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:145]: UNKNOWN is basically a stub for anything. For example, during test, we write a file using some fake key provider, and set it to unknown. As long as the reader uses the same key provider, it is able to read the file without making the fake provider an actual provider in the proto

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java [hunk 2]
```diff
@@ -548,6 +551,20 @@ private void advanceToNextStripe()
...
     }
 
+    private List<Slice> getDecryptionKeyMetadata(int currentStripe)
+    {
+        // if this stripe has encryption keys, then those are used
+        // otherwise look at nearest prior stripe that specifies encryption keys
+        // if the first stripe has no encryption information, then there are no encrypted columns;
+        for (int i = currentStripe; i >= 0; i--) {
+            if (!stripes.get(i).getKeyMetadata().isEmpty()) {
+                return stripes.get(i).getKeyMetadata();
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:561]: do we have check somewhere that checks key count matches number of encryption groups

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java [hunk 3]
```diff
@@ -159,7 +161,7 @@ public StripeFooter readStripeFooter(List<OrcType> types, InputStream inputStrea
     {
         CodedInputStream input = CodedInputStream.newInstance(inputStream);
         OrcProto.StripeFooter stripeFooter = OrcProto.StripeFooter.parseFrom(input);
-        return new StripeFooter(toStream(stripeFooter.getStreamsList()), toColumnEncoding(stripeFooter.getColumnsList()));
+        return new StripeFooter(toStream(stripeFooter.getStreamsList()), toColumnEncoding(stripeFooter.getColumnsList()), ImmutableList.of());
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java:164]: ditto

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java [hunk 4]
```diff
@@ -133,7 +133,8 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
                 toStripeInformation(footer.getStripesList()),
                 toType(footer.getTypesList()),
                 toColumnStatistics(hiveWriterVersion, footer.getStatisticsList(), false),
-                toUserMetadata(footer.getMetadataList()));
+                toUserMetadata(footer.getMetadataList()),
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java:136]: do you have pending commit to extract EncryptionGroup from orc footer and wire it up here?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 5]
```diff
@@ -105,7 +109,41 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
                 toType(footer.getTypesList()),
                 fileStats,
-                toUserMetadata(footer.getMetadataList()));
+                toUserMetadata(footer.getMetadataList()),
+                toEncryption(footer.getEncryption()));
+    }
+
+    private Optional<Encryption> toEncryption(DwrfProto.Encryption encryption)
+    {
+        if (encryption.getEncryptionGroupsCount() == 0) {
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:118]: the contract is that proto.encryption is optional. If it exists, it should have at least one encryption group

File: presto-orc/src/main/java/com/facebook/presto/orc/DwrfDecryptorProvider.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        return new DummyDecryptor();
+    }
+
+    private Decryptor getUnknownDecryptor()
+    {
+        // TODO: configure and pass through or throw an exception??
+        return new DummyDecryptor();
+    }
+
+    public Optional<DwrfDecryptor> createDecryptorIfEncrypted(int id, List<Slice> keyMetadatas)
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/DwrfDecryptorProvider.java:72]: in theory, decrypter is stateless, and can be shared across all nodes in the same group

File: presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java [hunk 7]
```diff
@@ -118,37 +124,31 @@ public StripeReader(
...
-                }
+        hasRowGroupDictionary = addIncludedStreams(columnEncodings, stripeFooter.getStreams(), streams);
+
+        // if streams doesn't have all included orcColumns, some included columns are encrypted
+        // get encrypted streams
+        if (streams.size() < includedOrcColumns.size()) {
+            verify(decryptorProvider.isPresent(), "decryptorProvider is absent");
+            List<Slice> encryptedStripeEncryptionGroups = stripeFooter.getStripeEncryptionGroups();
+            for (int i = 0; i < encryptedStripeEncryptionGroups.size(); i++) {
+                DwrfDecryptor decryptor = decryptorProvider.get().createDecryptor(decryptionKeyMetadata.get(i));
```
Comment (zzhao0) [inline presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java:147]: in theory, decrypter doesn't need to be recreated as long as key metadata doesn't change

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java (hunk 4 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java:136]: No, for now the plan is just to add support for encrypted DWRF, and not encrypted ORC.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 8]
```diff
@@ -129,12 +177,20 @@ public StripeFooter readStripeFooter(List<OrcType> types, InputStream inputStrea
     {
         CodedInputStream input = CodedInputStream.newInstance(inputStream);
         DwrfProto.StripeFooter stripeFooter = DwrfProto.StripeFooter.parseFrom(input);
-        return new StripeFooter(toStream(stripeFooter.getStreamsList()), toColumnEncoding(types, stripeFooter.getColumnsList()));
+        return new StripeFooter(toStream(
+                stripeFooter.getStreamsList()),
+                toColumnEncoding(types, stripeFooter.getColumnsList()),
+                stripeFooter.getEncryptedGroupsList().stream().map(OrcMetadataReader::byteStringToSlice).collect(toImmutableList()));
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:182]: nit: maybe wrap it

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+// of each are slightly different?
+public class Encryption
+{
+    private final KeyProvider keyProvider;
+    private final List<EncryptionGroup> encryptionGroups;
+
+    public Encryption(KeyProvider keyProvider, List<EncryptionGroup> encryptionGroups)
+    {
+        this.keyProvider = requireNonNull(keyProvider, "keyProvider is null");
+        this.encryptionGroups = ImmutableList.copyOf(encryptionGroups);
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java:32]: nit: requireNonNull

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Footer.java [hunk 10]
```diff
@@ -44,6 +46,7 @@ public Footer(long numberOfRows, int rowsInRowGroup, List<StripeInformation> str
         this.fileStats = ImmutableList.copyOf(requireNonNull(fileStats, "columnStatistics is null"));
         requireNonNull(userMetadata, "userMetadata is null");
         this.userMetadata = ImmutableMap.copyOf(transformValues(userMetadata, Slices::copyOf));
+        this.encryption = requireNonNull(encryption);
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/Footer.java:49]: nit: message

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.metadata;
+
+public enum KeyProvider
+{
+    UNKNOWN,
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java:18]: This does look weird. Do we want to drop it and throw an exception instead?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java [hunk 12]
```diff
@@ -41,18 +43,21 @@
...
 
-    public Stream(int column, StreamKind streamKind, int length, boolean useVInts, int sequence)
+    public Stream(int column, StreamKind streamKind, int length, boolean useVInts, int sequence, Optional<Long> offset)
     {
         this.column = column;
         this.streamKind = requireNonNull(streamKind, "streamKind is null");
         this.length = length;
         this.useVInts = useVInts;
         this.sequence = sequence;
+        this.offset = offset;
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/Stream.java:60]: `requireNonNull`

File: presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+package com.facebook.presto.orc;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.nio.ByteBuffer;
+import java.util.Base64;
+import java.util.Base64.Decoder;
+
+public class Base64Decoder
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java:23]: Maybe `Base64Decryptor`?

File: presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        implements Decryptor
+{
+    private final Decoder decoder = Base64.getDecoder();
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java:26]: static?

File: presto-orc/src/main/java/com/facebook/presto/orc/DwrfDecryptor.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import io.airlift.slice.Slice;
+
+public class DwrfDecryptor
+{
+    private final Slice keyMetadata;
+    private final Decryptor decryptor;
+
+    public DwrfDecryptor(Slice keyMetadata, Decryptor decryptor)
+    {
+        this.keyMetadata = keyMetadata;
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/DwrfDecryptor.java:25]: nit: requireNonNull for both

File: presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java (hunk 13 above)
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java:23]: As I understand it is just some stub implementation for testing, that actually does no encryption, just Base64 encoding, right?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/StripeFooter.java [hunk 16]
```diff
@@ -24,10 +25,14 @@
...
 
-    public StripeFooter(List<Stream> streams, List<ColumnEncoding> columnEncodings)
+    // encrypted StripeEncryptionGroups
+    private final List<Slice> stripeEncryptionGroups;
+
+    public StripeFooter(List<Stream> streams, List<ColumnEncoding> columnEncodings, List<Slice> stripeEncryptionGroups)
     {
         this.streams = ImmutableList.copyOf(requireNonNull(streams, "streams is null"));
         this.columnEncodings = ImmutableList.copyOf(requireNonNull(columnEncodings, "columnEncodings is null"));
+        this.stripeEncryptionGroups = requireNonNull(stripeEncryptionGroups, "stripeEncryptionGroups is null");
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/StripeFooter.java:35]: `ImmutableList.copyOf`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java [hunk 17]
```diff
@@ -127,6 +132,27 @@ public OrcReader(
         this.cacheable = requireNonNull(cacheable, "hiveFileContext is null");
     }
 
+    private static void validateEncryption(Footer footer, OrcDataSource dataSource)
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:135]: nit: This can simply take `OrcDataSourceId`

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java [hunk 18]
```diff
@@ -532,8 +539,14 @@ private void advanceToNextStripe()
 
         StripeInformation stripeInformation = stripes.get(currentStripe);
         validateWriteStripe(stripeInformation.getNumberOfRows());
+        List<Slice> stripeDecryptionKeyMetadata = getDecryptionKeyMetadata(currentStripe);
+        if ((!stripeDecryptionKeyMetadata.isEmpty() && !currentGroupDwrfDecryptors.isPresent())
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:543]: It is quite hard to follow this condition. What is checked here?

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java [hunk 19]
```diff
@@ -532,8 +539,14 @@ private void advanceToNextStripe()
...
+                || (currentGroupDwrfDecryptors.isPresent() && stripeDecryptionKeyMetadata.equals(currentGroupDwrfDecryptors.get().getKeyMetadatas()))) {
+            verify(decryptorProvider.isPresent(), "decryptorProvider is absent");
+            currentGroupDwrfDecryptors.equals(new GroupDwrfDecryptors(decryptorProvider.get(), stripeDecryptionKeyMetadata));
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:546]: `currentGroupDwrfDecryptors ` is `Optional`. This condition is always false.

Comment (arhimondr) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:546]: `currentGroupDwrfDecryptors ` is never set

File: presto-orc/src/main/java/com/facebook/presto/orc/DecryptingDecompressor.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+public class DecryptingDecompressor
+        implements OrcDecompressor
+{
+    private final DwrfDecryptor decryptor;
+    private final OrcDecompressor decompressor;
+
+    public DecryptingDecompressor(DwrfDecryptor decryptor, OrcDecompressor decompressor)
+    {
+        this.decompressor = decompressor;
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/DecryptingDecompressor.java:26]: nit: `requireNonNull`

File: presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java (hunk 13 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/Base64Decoder.java:23]: correct

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java (hunk 19 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:546]: ah sorry, that was supposed to be setting the value.

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java (hunk 18 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:543]: if there are encrytped columns and currentGroupDwrfDecryptors hasn't been set yet
or it has been set, but we have new encryption keys (that second condition was missing a `!`) 

Then we update currentGroupDwrfDecryptors for the new keys

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java (hunk 11 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java:18]: see https://github.com/prestodb/presto/pull/14483#discussion_r420431194

**Comment by rschlussel** [comment]
All comments addressed.

I've added unit tests for parts of the logic.  I'll test actually reading encrypted files via roundtrip testing when the writer is done.  I can fix any issues that arise then.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 21]
```diff
@@ -105,7 +109,38 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+        for (DwrfProto.EncryptionGroup dwrfEncryptionGroup : encryptionGroups) {
+            encryptionGroupBuilder.add(new EncryptionGroup(
+                    dwrfEncryptionGroup.getNodesList(),
+                    dwrfEncryptionGroup.hasKeyMetadata() ? Optional.of(byteStringToSlice(dwrfEncryptionGroup.getKeyMetadata())) : Optional.empty(),
+                    byteStringToSlice(dwrfEncryptionGroup.getStatistics())));
+        }
+        return encryptionGroupBuilder.build();
+    }
+
+    private KeyProvider toKeyProvider(DwrfProto.Encryption.KeyProvider keyProvider)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:135]: this method can be static

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 22]
```diff
@@ -105,7 +109,38 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
...
+    }
+
+    private Encryption toEncryption(DwrfProto.Encryption encryption)
+    {
+        KeyProvider keyProvider = toKeyProvider(encryption.getKeyProvider());
+        List<EncryptionGroup> encryptionGroups = toEncryptionGroups(encryption.getEncryptionGroupsList());
+        return new Encryption(keyProvider, encryptionGroups);
+    }
+
+    private List<EncryptionGroup> toEncryptionGroups(List<DwrfProto.EncryptionGroup> encryptionGroups)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:123]: this method can be static

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 23]
```diff
@@ -105,7 +109,38 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
                 toStripeInformation(footer.getStripesList()),
                 toType(footer.getTypesList()),
                 fileStats,
-                toUserMetadata(footer.getMetadataList()));
+                toUserMetadata(footer.getMetadataList()),
+                footer.hasEncryption() ? Optional.of(toEncryption(footer.getEncryption())) : Optional.empty());
+    }
+
+    private Encryption toEncryption(DwrfProto.Encryption encryption)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:116]: this method can be static

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java [hunk 24]
```diff
@@ -120,7 +154,17 @@ private static StripeInformation toStripeInformation(DwrfProto.StripeInformation
                 stripeInformation.getOffset(),
                 stripeInformation.getIndexLength(),
                 stripeInformation.getDataLength(),
-                stripeInformation.getFooterLength());
+                stripeInformation.getFooterLength(),
+                toKeyMetadatas(stripeInformation.getKeyMetadataList()));
+    }
+
+    private static List<Slice> toKeyMetadatas(List<ByteString> keyMetadataList)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:161]: any reason not to use stream APIs here and inlining like in readStripeFooter?

```
        return keyMetadataList.stream()
                .map(OrcMetadataReader::byteStringToSlice)
                .collect(toImmutableList());
```

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+
+public class EncryptionGroup
+{
+    // Sub trees in the schema that encryption should be applied
+    // Sub tree is identified by its root id
+    private final List<Integer> nodes;
+    // Arbitrary binary representing key metadata. It could be identifier
+    // of key in KMS, encrypted DEK or other form of user defined key metadata.
+    //  this is optional and can use first keyMetadata from stripe
+    private final Optional<Slice> keyMetadata;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java:30]: keyMetadata and statistics are not used? Why is that?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+    // encrypted columns stats (serialized FileStatistics)
+    private final Slice statistics;
+
+    public EncryptionGroup(List<Integer> nodes, Optional<Slice> keyMetadata, Slice statistics)
+    {
+        this.nodes = ImmutableList.copyOf(nodes);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java:36]: requireNotNull for all arguments?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java (hunk 11 above)
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/KeyProvider.java:18]: perhaps, add a comment explaining why/how UNKNOWN is used

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package com.facebook.presto.orc.metadata;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class Encryption
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java:22]: You mentioned somewhere that this class will be different for ORC, hence, adding an interface and Dwrf implementation would be appropriate.

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java (hunk 2 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:577]: @rschlussel What do you think about @zzhao0 's question here?

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java [hunk 28]
```diff
@@ -548,6 +551,20 @@ private void advanceToNextStripe()
...
+        // if this stripe has encryption keys, then those are used
+        // otherwise look at nearest prior stripe that specifies encryption keys
+        // if the first stripe has no encryption information, then there are no encrypted columns;
+        for (int i = currentStripe; i >= 0; i--) {
+            if (!stripes.get(i).getKeyMetadata().isEmpty()) {
+                return stripes.get(i).getKeyMetadata();
+            }
+        }
+
+        return ImmutableList.of();
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:581]: @rschlussel What do you think about @zzhao0 's question here?

[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:581]: I added checks for that when we read the footer in OrcReader (validateEncryption()

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java (hunk 2 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java:577]: I added checks for that when we read the footer in OrcReader (validateEncryption()

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java (hunk 25 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/EncryptionGroup.java:30]: the key is the one that encrypts the statistics.  We don't use statistics because of https://github.com/prestodb/presto/blob/561e977f3b39c056ef76ea00ff826f0148039b93/presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java#L98

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java (hunk 27 above)
Comment (rschlussel) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java:22]: The reason I decided against it is that there are a number of differences, and it's not clear what would go in a common interface.  I think it's the right decision long term, but I figure it should wait for the ORC implementation to figure out what that interface should look like (is it empty and we just cast to the appropriate encryption type for each, or are ther common methods?).  what if I rename this DwrfEncryption so that it's clear it's separate from the OrcEncryption?

Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/Encryption.java]: @rschlussel 

> what if I rename this DwrfEncryption so that it's clear it's separate from the OrcEncryption?

Good idea. Let's do that.

**Comment by rschlussel** [comment]
@mbasmanova thanks so much for the review!  I realized that when I updated last night to separate something out of the last commit, I left all the new files out of the last commit. I'm not sure if you saw those or not while you were reviewing. They should be back in the commit now

**Comment by rschlussel** [comment]
closing in favor of #14621, which includes some bug fixes to this PR
